vissl/losses/barlow_twins_loss.py

```diff
@@ -0,0 +1,211 @@
+    https://arxiv.org/abs/2103.03230v1.
+
+    Args:
+        lambda_ (float):        weight on the off-diagonal terms. It controls the trade-off
+                                between the importance given to the invariance term versus the
+                                redundancy reduction term.
+        scale_loss (float):     In order to match the code that was used to develop Barlow
+                                Twins, we include an additional parameter, scale-loss, that
+                                multiplies the loss by a constant factor. We are working on a
+                                version that will not require this parameter.
+        embedding_dim (int):    dimensionality of the representation
+    """
+
+    def __init__(self, lambda_: float, scale_loss: float, embedding_dim: int):
+        super(BarlowTwinsCriterion, self).__init__()
+
+        self.lambda_ = lambda_
+        self.scale_loss = scale_loss
+        self.embedding_dim = embedding_dim
+        self.num_copies = 2
+        self.eps = 1e-5
+
+    @staticmethod
+    def _off_diagonal(x: torch.Tensor) -> torch.Tensor:
+        """
+        return a flattened view of the off-diagonal elements of a square matrix
+       """
+        n, m = x.shape
+        assert n == m
+        return x.flatten()[:-1].view(n - 1, n + 1)[:, 1:].flatten()
+
+    def forward(self, embedding: torch.Tensor) -> torch.tensor:
+        """
+        Calculate the loss. Operates on embeddings tensor.
+
+        Args:
+            embedding (torch.Tensor):   NxEMBEDDING_DIM
+                                        Must contain the concatenated embeddings
+                                        of the two image copies:
+                                        [emb_img1_0, emb_img2_0, ....., emb_img1_1, emb_img2_1,...]
+        """
+        assert embedding.ndim == 2 and embedding.shape[1] == int(
+            self.embedding_dim
+        ), f"Incorrect embedding shape: {embedding.shape} but expected Nx{self.embedding_dim}"
+
+        batch_size = embedding.shape[0]
+        assert (
+                batch_size % self.num_copies == 0
+        ), f"Batch size {batch_size} should be divisible by num_copies ({self.num_copies})."
+
+        # normalize embeddings along the batch dimension
+        embedding_normed = SyncNormalizeFunction.apply(embedding, self.eps)
+
+        # split embedding between copies
+        embedding_normed_a, embedding_normed_b = torch.split(
+            embedding_normed,
+            split_size_or_sections=batch_size // self.num_copies,
+            dim=0,
+        )
+
+        # Do a gather over all embeddings, so we can compute the loss.
+        # Final shape is: (batch_size * num_gpus) x embedding_dim
+        embedding_normed_a = gather_from_all(embedding_normed_a)
+        embedding_normed_b = gather_from_all(embedding_normed_b)
+
+        # cross-correlation matrix
+        c = torch.mm(embedding_normed_a.T, embedding_normed_b) / (
```

Comment:
nit: let's use `c` -> `correlation_matrix` something like that for more verbosity :)